sklearn/decomposition/nmf.py

```diff
@@ -254,22 +343,285 @@ def _nls_subproblem(V, W, H, tol, max_iter, sigma=0.01, beta=0.1):
+                         * np.ones((1, n_components_))]),
+            H, tolH, nls_max_iter, regul=alpha, l1_ratio=l1_ratio)
+
+    return H, gradH, iterH
+
+
+def _fit_projected_gradient(X, W, H, tol, max_iter,
+                            nls_max_iter, alpha, l1_ratio,
+                            sparseness, beta, eta):
+    """Compute Non-negative Matrix Factorization (NMF) with Projected Gradient
+    """
+    gradW = (np.dot(W, np.dot(H, H.T))
+             - safe_sparse_dot(X, H.T, dense_output=True))
+    gradH = (np.dot(np.dot(W.T, W), H)
+             - safe_sparse_dot(W.T, X, dense_output=True))
+
+    init_grad = squared_norm(gradW) + squared_norm(gradH.T)
+    # max(0.001, tol) to force alternating minimizations of W and H
+    tolW = max(0.001, tol) * np.sqrt(init_grad)
+    tolH = tolW
 
-    Read more in the :ref:`User Guide <NMF>`.
+    for n_iter in range(1, max_iter + 1):
+        # stopping condition
+        # as discussed in paper
+        proj_grad_W = squared_norm(gradW * np.logical_or(gradW < 0, W > 0))
+        proj_grad_H = squared_norm(gradH * np.logical_or(gradH < 0, H > 0))
+
+        if (proj_grad_W + proj_grad_H) / init_grad < tol ** 2:
+            break
+
+        # update W
+        W, gradW, iterW = _update_projected_gradient_w(X, W, H, tolW,
+                                                       nls_max_iter,
+                                                       alpha, l1_ratio,
+                                                       sparseness, beta, eta)
+        if iterW == 1:
+            tolW = 0.1 * tolW
+
+        # update H
+        H, gradH, iterH = _update_projected_gradient_h(X, W, H, tolH,
+                                                       nls_max_iter,
+                                                       alpha, l1_ratio,
+                                                       sparseness, beta, eta)
+        if iterH == 1:
+            tolH = 0.1 * tolH
+
+    H[H == 0] = 0   # fix up negative zeros
+
+    if n_iter == max_iter:
+        warnings.warn("Iteration limit reached during fit. ")
+        W, _, _ = _update_projected_gradient_w(X, W, H, tol, nls_max_iter,
+                                               alpha, l1_ratio, sparseness,
+                                               beta, eta)
+
+    return W, H, n_iter
+
+
+def _update_coordinate_descent(X, W, Ht, alpha, l1_ratio, tol_greedy, greedy):
+    """Helper function for _fit_coordinate_descent"""
+    n_components = Ht.shape[1]
+
+    # L1 and L2 regularization
+    l1_reg = l1_ratio * alpha
+    l2_reg = (1 - l1_ratio) * alpha
+
+    # L2 regularization corresponds to increase the diagonal of HHt
+    HHt = np.dot(Ht.T, Ht) + np.eye(n_components) * l2_reg
+    # L1 regularization correponds to decrease each element of HX
+    XHt = safe_sparse_dot(X, Ht) - l1_reg
+
+    # gives C-contiguous array to Cython
+    if HHt.flags['F_CONTIGUOUS']:
+        HHt = HHt.T
+
+    if not greedy:
```

Comment:
Don't know if this will stay in, but I think it's conventional to have the positive branch of the if/else first if it makes no difference.